fvm/transactionStorageLimiter.go

```diff
@@ -14,29 +14,73 @@ import (
 
 type TransactionStorageLimiter struct{}
 
-func (_ TransactionStorageLimiter) CheckStorageLimits(
+// CheckStorageLimits checks each account that had its storage written to during a transaction, that its storage used
+// is less than its storage capacity.
+// Storage used is an FVM register and is easily accessible.
+// Storage capacity is calculated by the FlowStorageFees contract from the account's flow balance.
+//
+// The payers balance is considered to be maxTxFees lower that its actual balance, due to the fact that
+// the fee deduction step happens after the storage limit check.
+func (d TransactionStorageLimiter) CheckStorageLimits(
 	env environment.Environment,
 	addresses []flow.Address,
+	payer flow.Address,
+	maxTxFees uint64,
 ) error {
 	if !env.LimitAccountStorage() {
 		return nil
 	}
 
 	defer env.StartSpanFromRoot(trace.FVMTransactionStorageUsedCheck).End()
 
+	err := d.checkStorageLimits(env, addresses, payer, maxTxFees)
+	if err != nil {
+		return fmt.Errorf("storage limit check failed: %w", err)
+	}
+	return nil
+}
+
+func (d TransactionStorageLimiter) checkStorageLimits(
+	env environment.Environment,
+	addresses []flow.Address,
+	payer flow.Address,
+	maxTxFees uint64,
+) error {
+	// in case the payer is not already part of the check, include it here.
+	// If the maxTxFees is zero, it doesn't matter if the payer is included or not.
+	if maxTxFees > 0 {
+		commonAddressesContainPayer := false
+		for _, address := range addresses {
+			if address == payer {
+				commonAddressesContainPayer = true
+				break
+			}
+		}
+
+		if !commonAddressesContainPayer {
+			addresses = append(addresses, payer)
+		}
+	}
+
 	commonAddresses := make([]common.Address, len(addresses))
 	usages := make([]uint64, len(commonAddresses))
+
 	for i, address := range addresses {
-		c := common.Address(address)
-		commonAddresses[i] = c
-		u, err := env.GetStorageUsed(c)
+		ca := common.Address(address)
+		u, err := env.GetStorageUsed(ca)
 		if err != nil {
-			return fmt.Errorf("storage limit check failed: %w", err)
+			return err
 		}
+
+		commonAddresses[i] = ca
 		usages[i] = u
 	}
 
-	result, invokeErr := env.AccountsStorageCapacity(commonAddresses)
+	result, invokeErr := env.AccountsStorageCapacity(
+		commonAddresses,
+		common.Address(payer),
+		maxTxFees,
```

Comment:
Is `AccountsStorageCapacity` actually `CheckMaxTxFeeCanBeCoveredByPayerBalanceAndReadAccountsStorageCapacity`? Meaning it performs 1 check and 1 read:
1. check the payer has enough balance to cover maxTxFees
2. read each account's storage capacity

If so, why don't we separate into two calls instead?